gtsfm/evaluation/visualize_benchmark_comparison.py

```diff
@@ -0,0 +1,254 @@
+
+        # loop over each benchmark result (columns of table)
+        for zip_fname in ZIP_FNAMES:
+            # use just the first 35 chars
+            X.append(zip_fname[:MAX_NUM_CHARS_ARTIFACT_FNAME])
+
+            report1_fpath = f"{curr_master_dirpath}/results-{zip_fname}/result_metrics/gtsfm_metrics_report.html"
+            tables_dict1 = report_utils.extract_tables_from_report(report1_fpath)
+
+            report2_fpath = f"{new_branch_dirpath}/results-{zip_fname}/result_metrics/gtsfm_metrics_report.html"
+            tables_dict2 = report_utils.extract_tables_from_report(report2_fpath)
+            merged_tables_dict = report_utils.merge_tables(tables_dict1, tables_dict2)
+
+            # loop over each metric within this table (rows of table)
+            for i, (metric_name, master_val, branch_val) in enumerate(merged_tables_dict[table_name]):
+
+                if branch_val is None:
+                    percentage_change = np.nan
+                else:
+                    percentage_change = compute_percentage_change(float(master_val), float(branch_val))
+
+                if "error" in metric_name and "outlier" not in metric_name:
+                    # smaller is better, so this will flip the color to green for reduced values, instead of red
+                    # exception are outlier errors, which we want to get larger.
+                    percentage_change *= -1
+                benchmark_table_vals[metric_name][zip_fname] = percentage_change
+
+        Z_rows = []
+        for metric_name, benchmark_vals_dict in benchmark_table_vals.items():
+            Z_row = []
+            for zip_fname in ZIP_FNAMES:
+                Z_row.append(benchmark_vals_dict.get(zip_fname, np.nan))  # default was unchanged if missing
+            Z_rows.append(Z_row)
+            Y.append(metric_name)
+
+        Z = np.array(Z_rows)
+        html = plot_colored_table(X, Y, Z, table_name=table_name)
+
+        # Write name of the metric group in human readable form.
+        f.write(metrics_report.get_html_metric_heading(table_name))
+        f.write(html)
+
+        # Close HTML tags.
+        f.write("</html>")
+
+
+def compute_percentage_change(x: float, y: float) -> float:
+    """Return percentage in representing the regression or improvement of a value x, for new value y."""
+    return (y - x) / (x + EPSILON) * 100
+
+
+def test_compute_percentage_change_improve() -> None:
+    """ """
+    x = 100
+    y = 150
+    change_percent = compute_percentage_change(x, y)
+    assert np.isclose(change_percent, 50)
+
+
+def test_compute_percentage_change_static() -> None:
+    """ """
+    x = 100
+    y = 100
+    change_percent = compute_percentage_change(x, y)
+    assert np.isclose(change_percent, 0)
+
+
+def test_compute_percentage_change_regression() -> None:
+    """ """
+    x = 100
+    y = 1
+    change_percent = compute_percentage_change(x, y)
+    assert np.isclose(change_percent, -99)
```

Comment:
why are these tests here? I am surprised to see so many tests for a simple formula. If you really need them, can we move the function to a util and the tests to its tests? I think there is a util for evaluation or metrics

Reply:
sure, i moved them to utils/metrics.py and the corresponding unit test file now

Reply:
I'm still seeing three tests in this file.